### PR TITLE
Cleanup DiscordUser#GetAvatarUrl and add DiscordMember#GetGuildAvatarUrl

### LINKED ISSUE #1146: DiscordMembers should have a method like GetAvatarUrl, possibly GetGuildAvatarUrl
# Summary
I think the title's pretty self explanatory. I could probably implement this myself but if somebody gets to it before me I won't stop them, I wanted to create this issue before the PR so somebody can do that if I'm not able to make a PR in time (and for opinions, of course).

# Details
It would be nice to be able to have the same functionality as choosing the image format and/or changing the size of the returned URL just like how it's done with `GuildUser#GetAvatarUrl`.

### PR DESCRIPTION
**Pull request by OoLunar**
# Summary
Cleans up the DiscordUser#GetAvatarUrl method by renaming variables, adding code comments and using simplified logic. Adds DiscordMember#GetGuildAvatarUrl by copying DiscordUser#GetAvatarUrl with slight modifications, including falling back to DiscordUser#GetAvatarUrl when no `GuildAvatarHash` is set.

# Details
Additionally simplfies a few return if statements and fixes an XML doc.

Testing guild avatars:
```cs
using System.Threading.Tasks;
using DSharpPlus;
using DSharpPlus.Entities;
using DSharpPlus.SlashCommands;

namespace Tomoe.Commands
{
    public partial class Test : ApplicationCommandModule
    {
        [SlashCommand("guild_avatar_test", "Tests if D#+ can grab the members avatar at 4096px")]
        public static async Task GuildAvatarTest(InteractionContext context)
        {
            DiscordInteractionResponseBuilder messageBuilder = new();
            messageBuilder.AddEmbed(new DiscordEmbedBuilder()
            {
                Title = "Guild Avatar Test",
                Description = "This is a test of the member's guild avatar. If you see this, it means that D#+ can grab the member's guild avatar at 2048px.",
                ImageUrl = context.Member.GetGuildAvatarUrl(ImageFormat.Png, 2048)
            });

            messageBuilder.AddEmbed(new DiscordEmbedBuilder()
            {
                Title = "Guild Avatar Test",
                Description = "This is a test of the member's guild avatar. If you see this, it means D#+ can grab the member's guild avatar at 4096px.",
                ImageUrl = context.Member.GetGuildAvatarUrl(ImageFormat.Png, 4096)
            });

            await context.CreateResponseAsync(InteractionResponseType.ChannelMessageWithSource, messageBuilder);
        }
    }
}
```
Testing guild avatars:
image

Testing user avatars:
```cs
using System.Threading.Tasks;
using DSharpPlus;
using DSharpPlus.Entities;
using DSharpPlus.SlashCommands;

namespace Tomoe.Commands
{
    public partial class Test : ApplicationCommandModule
    {
        [SlashCommand("user_avatar_test", "Tests if D#+ can grab the user's avatar at 4096px.")]
        public static async Task UserAvatarTest(InteractionContext context)
        {
            DiscordInteractionResponseBuilder messageBuilder = new();
            messageBuilder.AddEmbed(new DiscordEmbedBuilder()
            {
                Title = "User Avatar Test",
                Description = "This is a test of the user's avatar. If you see this, it means that D#+ can grab the user's avatar at 2048px.",
                ImageUrl = context.User.GetAvatarUrl(ImageFormat.Png, 2048)
            });

            messageBuilder.AddEmbed(new DiscordEmbedBuilder()
            {
                Title = "User Avatar Test",
                Description = "This is a test of the user's avatar. If you see this, it means D#+ can grab the user's avatar at 4096px.",
                ImageUrl = context.User.GetAvatarUrl(ImageFormat.Png, 4096)
            });

            await context.CreateResponseAsync(InteractionResponseType.ChannelMessageWithSource, messageBuilder);
        }
    }
}
```
image

Closes #1146

### CODE REVIEW

**Comment by VelvetThePanda** [comment]
> including falling back to DiscordUser#GetAvatarUrl when no GuildAvatarHash is set.

This was actualyl a complaint about the current implementation of guild avatars; I'd advise against this.

File: DSharpPlus/Entities/Guild/DiscordMember.cs [hunk 1]
```diff
@@ -587,6 +587,45 @@ public async Task UpdateVoiceStateAsync(DiscordChannel channel, bool? suppress)
...
+        /// <param name="imageFormat">The image format of the avatar to get.</param>
+        /// <param name="imageSize">The maximum size of the avatar. Must be a power of two, minimum 16, maximum 4096.</param>
+        /// <returns>The URL of the user's avatar.</returns>
+        public string GetGuildAvatarUrl(ImageFormat imageFormat, ushort imageSize = 1024)
+        {
+            // Run this if statement before any others to prevent running the if statements twice.
+            if (string.IsNullOrWhiteSpace(this.AvatarHash))
+            {
+                return this.GetAvatarUrl(imageFormat, imageSize);
+            }
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Guild/DiscordMember.cs:602]: Nitpick: Braces can be elided here.

File: DSharpPlus/Entities/Guild/DiscordMember.cs [hunk 2]
```diff
@@ -587,6 +587,45 @@ public async Task UpdateVoiceStateAsync(DiscordChannel channel, bool? suppress)
...
+
+            if (imageFormat == ImageFormat.Unknown)
+                throw new ArgumentException("You must specify valid image format.", nameof(imageFormat));
+
+            // Makes sure the image size is in between Discord's allowed range.
+            if (imageSize < 16 || imageSize > 4096)
+                throw new ArgumentOutOfRangeException("Image Size is not in between 16 and 4096: " + nameof(imageSize));
+
+            // Checks to see if the image size is a power of two.
+            if (imageSize is 0 && (imageSize & (imageSize - 1)) is not 0)
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/Guild/DiscordMember.cs:612]: I did say `if (x is *not* 0`

**Comment by OoLunar** [comment]
> > including falling back to DiscordUser#GetAvatarUrl when no GuildAvatarHash is set.
> 
> This was actualyl a complaint about the current implementation of guild avatars; I'd advise against this.

That's a complaint? What should the behavior be instead, returning null?

File: DSharpPlus/Entities/User/DiscordUser.cs [hunk 3]
```diff
@@ -23,6 +23,7 @@
 
 using System;
 using System.Collections.Generic;
+using System.Dynamic;
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/User/DiscordUser.cs:26]: Concern.

**Review by VelvetThePanda — CHANGES_REQUESTED** [review]

**Comment by VelvetThePanda** [comment]
> > > including falling back to DiscordUser#GetAvatarUrl when no GuildAvatarHash is set.
> > 
> > 
> > This was actualyl a complaint about the current implementation of guild avatars; I'd advise against this.
> 
> That's a complaint? What should the behavior be instead, returning null?

Yes. There's also an issue with `DiscordMember.GuildAvatarHash`, or perhaps its with `GuildAvatarUrl`, where it falls back to the user avatar, if you'd like to fix that as well.

File: DSharpPlus/Entities/Guild/DiscordMember.cs (hunk 2 above)
Comment (OoLunar) [reply DSharpPlus/Entities/Guild/DiscordMember.cs:612]: Yea, we doing checks to make sure it is a power of two, and throw an exception if it isn't. Misleading comment, will fix

File: DSharpPlus/Entities/User/DiscordUser.cs (hunk 3 above)
Comment (OoLunar) [reply DSharpPlus/Entities/User/DiscordUser.cs:26]: Unused import, not sure how that got there

File: DSharpPlus/Entities/Guild/DiscordMember.cs (hunk 2 above)
Comment (VelvetThePanda) [reply DSharpPlus/Entities/Guild/DiscordMember.cs:612]: No, the if check is wrong. As discussed on Discord I showed you this: `x is not 0 && x & (x - 1) is 0`. The `is` checks are backward.

**Comment by OoLunar** [comment]
> > > > including falling back to DiscordUser#GetAvatarUrl when no GuildAvatarHash is set.
> > > 
> > > 
> > > This was actualyl a complaint about the current implementation of guild avatars; I'd advise against this.
> > 
> > 
> > That's a complaint? What should the behavior be instead, returning null?
> 
> Yes. There's also an issue with `DiscordMember.GuildAvatarHash`, or perhaps its with `GuildAvatarUrl`, where it falls back to the user avatar, if you'd like to fix that as well.

~~It's not a bug, it's a feature!~~ Will do

**Comment by OoLunar** [comment]
Latest commits tested and confirmed to be functioning correctly:
image
image

File: DSharpPlus/Entities/Guild/DiscordMember.cs (hunk 2 above)
Comment (VelvetThePanda) [reply DSharpPlus/Entities/Guild/DiscordMember.cs:609]: This is still broken. by changing `is not 0` to `is 0`, you've created an unreachable condition (which the compiler may or may not warn you about), and this is never true, so an exception is never thrown. Furthermore, by inverting the checks, you've broken the entire computation. What should be done instead is wrapping this in parentheses and inverting the result of that.

**Review by VelvetThePanda — CHANGES_REQUESTED** [review]

**Comment by VelvetThePanda** [comment]
This was discussed on Discord, but here's what the proper Pow2 check looks like: 
image


And this is the result of the inverted check (which doesn't invert the logic, as I said before)
image

File: DSharpPlus/Entities/Guild/DiscordMember.cs (hunk 2 above)
Comment (OoLunar) [reply DSharpPlus/Entities/Guild/DiscordMember.cs]: Should be fixed on latest commit

File: DSharpPlus/Entities/User/DiscordUser.cs [hunk 4]
```diff
@@ -205,41 +206,45 @@ public DiscordPresence Presence
...
-                throw new ArgumentOutOfRangeException(nameof(size));
+            // Makes sure the image size is in between Discord's allowed range.
+            if (imageSize < 16 || imageSize > 4096)
+                throw new ArgumentOutOfRangeException("Image Size is not in between 16 and 4096: " + nameof(imageSize));
 
-            var log = Math.Log(size, 2);
-            if (log < 4 || log > 11 || log % 1 != 0)
-                throw new ArgumentOutOfRangeException(nameof(size));
+            // Checks to see if the image size is not a power of two.
+            if (!(imageSize is not 0 && (imageSize & (imageSize - 1)) is 0))
```
Comment (VelvetThePanda) [inline DSharpPlus/Entities/User/DiscordUser.cs:222]: Much better :)

**Review by VelvetThePanda — APPROVED** [review]
LGTM. Thanks!

**Comment by FireController1847** [comment]
Thanks for this! Sorry I wasn't able to get around to doing it, but glad someone else was :)

**Comment by OoLunar** [comment]
> Thanks for this! Sorry I wasn't able to get around to doing it, but glad someone else was :)

Not a problem! I'm just making my way around unimplemented features. I apologize for forgetting to ask if anyone else was wanting to implement them :pray:

**Pull request merged**